Dockerfile

```diff
@@ -87,10 +87,28 @@ WORKDIR /app
 COPY --from=backend /build/ .
 COPY --from=frontend /build/ wwwroot/build/
 
+# set capability of NET_BIND_SERVICE to allow containers to bind and expose privileged ports
+# set net_bind both [e]ffective & [p]ermitted
+RUN apt install libcap2 libcap2-bin -y
+# it is the dotnet binary that requires the capability to launch a kestrel server on privileged ports
+RUN setcap CAP_NET_BIND_SERVICE=+ep /usr/share/dotnet/dotnet
+
+# expose privileged ports
 EXPOSE 80
 EXPOSE 443
 EXPOSE 11111
 
+# add a group and a user > 10000 to avoid conflicts with system accounts
+RUN groupadd --gid 10000 --system runner
```

Comment:
Why is something like this not part of the base image from Microsoft?

Reply:
This is something Microsoft seem to be aware of and they have been very recently (Aug 22) making new images that are more secure by default like the recent `Ubuntu-Chiseled` containers but sadly these non-root changes seem to have only been made to Ubuntu and the likes of Alpine and Debian have currently been ignored.

https://devblogs.microsoft.com/dotnet/dotnet-6-is-now-in-ubuntu-2204/#non-root-images
https://github.com/dotnet/dotnet-docker/blob/fd39ea3eece8c8653348a717446ac41e93633784/src/runtime-deps/6.0/jammy-chiseled/arm64v8/Dockerfile#L13-L26

My guess would be that they are just slightly behind on container security best practices and will slowly adopt them into their offerings.

Reply:
To be honest I have never heard about it. It does not mean anything because I work in small startups where every developer is superadmin anyway ;) ... but even from my large customers with 100+ developers I have no heard about this before.

Reply:
I am not sure if it is just a blind-spot in the industry or if the concept is immature but it certainly needs more visibility.

Containers are just small linux instances capable of doing almost anything a regular linux machine is capable of, so when we allow containers to run as root we allow attackers a lot of room to escalate their attacks into the cluster and potentially even further.

For example, should an attacker find a vulnerability in an application that allows them to execute a shell in a container that has root access, they could use `apt` to install `curl`, `git`, `kubctl`, etc. They could then download malicious payloads into the container, enumerate other containers/pods in the cluster and widen their attack internally to find your persistent data stores where your sensitive data is stored, try to take down parts of the cluster, etc.

Sure, whilst such an attack would require another vulnerability to exist allowing the attacker to execute commands inside the container, we should never assume that this is not possible. The principal of "security by design" is to always expect attacks and an attacker is not going to stop once they have a foothold in a system, they will try to escalate that attack as far and wide as possible. We need to ensure that every layer of the onion is secure by design, and container security is an additional layer that needs to be secured.

An interesting read that has parallels to this topic:
https://www.cyberark.com/resources/threat-research-blog/how-i-hacked-play-with-docker-and-remotely-ran-code-on-the-host

The changes I am making in this PR, whilst mainly aimed at Kubernetes clusters and helm charts, are facilitated by the changes to the container image `Dockerfile` and with the appropriate supplementary changes, could be used to secure other deployment methods like `docker compose`, `docker swarm`, etc.